File: clients/go/protocol_test.go
package sixsevendb

import (
	"encoding/binary"
	"testing"
)

func TestBuildStartupMessage(t *testing.T) {
	msg := buildStartupMessage("testuser", "testdb")

	// First 4 bytes: length
	length := binary.BigEndian.Uint32(msg[0:4])
	if int(length) != len(msg) {
		t.Errorf("startup message length mismatch: header says %d, actual %d", length, len(msg))
	}

	// Next 4 bytes: protocol version
	version := binary.BigEndian.Uint32(msg[4:8])
	if version != protocolVersion {
		t.Errorf("protocol version = %d, want %d", version, protocolVersion)
	}

	// Params should contain user and database
	params := string(msg[8:])
	if !contains(params, "user\x00testuser\x00") {
		t.Error("startup message missing user parameter")
	}
	if !contains(params, "database\x00testdb\x00") {
		t.Error("startup message missing database parameter")
	}
}

func TestBuildQueryMessage(t *testing.T) {
	msg := buildQueryMessage("SELECT 1")

	if msg[0] != 'Q' {
		t.Errorf("query message type = %c, want Q", msg[0])
	}

	length := binary.BigEndian.Uint32(msg[1:5])
	expectedLen := uint32(4 + len("SELECT 1") + 1) // length field + sql + null terminator
	if length != expectedLen {
		t.Errorf("query message length = %d, want %d", length, expectedLen)
	}

	// SQL text + null terminator
	sqlText := string(msg[5 : len(msg)-1])
	if sqlText != "SELECT 1" {
		t.Errorf("query text = %q, want %q", sqlText, "SELECT 1")
	}
	if msg[len(msg)-1] != 0 {
		t.Error("query message missing null terminator")
	}
}

func TestBuildPasswordMessage(t *testing.T) {
	msg := buildPasswordMessage("secret")

	if msg[0] != 'p' {
		t.Errorf("password message type = %c, want p", msg[0])
	}
}

func TestBuildParseMessage(t *testing.T) {
	msg := buildParseMessage("SELECT $1", "")

	if msg[0] != 'P' {
		t.Errorf("parse message type = %c, want P", msg[0])
	}
}

func TestBuildBindMessage(t *testing.T) {
	msg := buildBindMessage([]string{"hello", "42"}, "", "")

	if msg[0] != 'B' {
		t.Errorf("bind message type = %c, want B", msg[0])
	}
}

func TestBuildDescribeMessage(t *testing.T) {
	msg := buildDescribeMessage('P', "")

	if msg[0] != 'D' {
		t.Errorf("describe message type = %c, want D", msg[0])
	}
}

func TestBuildExecuteMessage(t *testing.T) {
	msg := buildExecuteMessage("", 0)

	if msg[0] != 'E' {
		t.Errorf("execute message type = %c, want E", msg[0])
	}
}

func TestBuildSyncMessage(t *testing.T) {
	msg := buildSyncMessage()

	if msg[0] != 'S' {
		t.Errorf("sync message type = %c, want S", msg[0])
	}
	if len(msg) != 5 {
		t.Errorf("sync message length = %d, want 5", len(msg))
	}
}

func TestBuildTerminateMessage(t *testing.T) {
	msg := buildTerminateMessage()

	if msg[0] != 'X' {
		t.Errorf("terminate message type = %c, want X", msg[0])
	}
	if len(msg) != 5 {
		t.Errorf("terminate message length = %d, want 5", len(msg))
	}
}

func TestBuildSASLInitialResponse(t *testing.T) {
	msg := buildSASLInitialResponse("SCRAM-SHA-256", []byte("n,,n=user,r=nonce"))

	if msg[0] != 'p' {
		t.Errorf("SASL initial response type = %c, want p", msg[0])
	}
}

func TestBuildSASLResponse(t *testing.T) {
	msg := buildSASLResponse([]byte("c=biws,r=nonce,p=proof"))

	if msg[0] != 'p' {
		t.Errorf("SASL response type = %c, want p", msg[0])
	}
}

func TestParseRowDescription(t *testing.T) {
	// Build a synthetic RowDescription payload
	var payload []byte
	// Field count = 2
	payload = binary.BigEndian.AppendUint16(payload, 2)

	// Field 1: "id"
	payload = append(payload, []byte("id")...)
	payload = append(payload, 0) // null terminator
	payload = binary.BigEndian.AppendUint32(payload, 0)     // table OID
	payload = binary.BigEndian.AppendUint16(payload, 1)     // column index
	payload = binary.BigEndian.AppendUint32(payload, OIDInt4) // type OID
	payload = append(payload, 0, 4)                          // type size (int16 = 4)
	payload = binary.BigEndian.AppendUint32(payload, 0xFFFFFFFF) // type modifier -1
	payload = binary.BigEndian.AppendUint16(payload, 0)     // format code

	// Field 2: "name"
	payload = append(payload, []byte("name")...)
	payload = append(payload, 0)
	payload = binary.BigEndian.AppendUint32(payload, 0)
	payload = binary.BigEndian.AppendUint16(payload, 2)
	payload = binary.BigEndian.AppendUint32(payload, OIDText)
	payload = append(payload, 0xFF, 0xFF) // type size -1
	payload = binary.BigEndian.AppendUint32(payload, 0xFFFFFFFF)
	payload = binary.BigEndian.AppendUint16(payload, 0)

	fields, err := parseRowDescription(payload)
	if err != nil {
		t.Fatalf("parseRowDescription error: %v", err)
	}
	if len(fields) != 2 {
		t.Fatalf("expected 2 fields, got %d", len(fields))
	}
	if fields[0].Name != "id" {
		t.Errorf("field 0 name = %q, want %q", fields[0].Name, "id")
	}
	if fields[0].TypeOID != OIDInt4 {
		t.Errorf("field 0 type OID = %d, want %d", fields[0].TypeOID, OIDInt4)
	}
	if fields[1].Name != "name" {
		t.Errorf("field 1 name = %q, want %q", fields[1].Name, "name")
	}
	if fields[1].TypeOID != OIDText {
		t.Errorf("field 1 type OID = %d, want %d", fields[1].TypeOID, OIDText)
	}
}

func TestParseDataRow(t *testing.T) {
	var payload []byte
	// 2 columns
	payload = binary.BigEndian.AppendUint16(payload, 2)
	// Column 1: "42"
	payload = binary.BigEndian.AppendUint32(payload, 2)
	payload = append(payload, []byte("42")...)
	// Column 2: NULL
	payload = binary.BigEndian.AppendUint32(payload, 0xFFFFFFFF)

	values, err := parseDataRow(payload)
	if err != nil {
		t.Fatalf("parseDataRow error: %v", err)
	}
	if len(values) != 2 {
		t.Fatalf("expected 2 values, got %d", len(values))
	}
	if string(values[0]) != "42" {
		t.Errorf("column 0 = %q, want %q", string(values[0]), "42")
	}
	if values[1] != nil {
		t.Errorf("column 1 = %v, want nil", values[1])
	}
}

func TestParseErrorFields(t *testing.T) {
	payload := []byte("SERROR\x00C42000\x00Msyntax error\x00\x00")
	fields := parseErrorFields(payload)

	if fields['S'] != "ERROR" {
		t.Errorf("severity = %q, want %q", fields['S'], "ERROR")
	}
	if fields['C'] != "42000" {
		t.Errorf("code = %q, want %q", fields['C'], "42000")
	}
	if fields['M'] != "syntax error" {
		t.Errorf("message = %q, want %q", fields['M'], "syntax error")
	}
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && containsHelper(s, substr))
}

func containsHelper(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

File: clients/java/src/test/java/com/sixsevendb/ProtocolTest.java
package com.sixsevendb;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class ProtocolTest {

    @Test
    void testBuildStartupMessage() {
        byte[] msg = Protocol.buildStartupMessage("testuser", "testdb");
        assertNotNull(msg);
        assertTrue(msg.length > 8);
        // First 4 bytes: length, next 4 bytes: protocol version
        ByteBuffer buf = ByteBuffer.wrap(msg);
        int length = buf.getInt();
        assertEquals(msg.length, length);
        int version = buf.getInt();
        assertEquals(Protocol.PROTOCOL_VERSION, version);
        // Rest should contain "user\0testuser\0database\0testdb\0\0"
        String params = new String(msg, 8, msg.length - 8, StandardCharsets.UTF_8);
        assertTrue(params.contains("user"));
        assertTrue(params.contains("testuser"));
        assertTrue(params.contains("database"));
        assertTrue(params.contains("testdb"));
    }

    @Test
    void testBuildPasswordMessage() {
        byte[] msg = Protocol.buildPasswordMessage("secret");
        assertEquals('p', msg[0]);
        ByteBuffer buf = ByteBuffer.wrap(msg, 1, 4);
        int length = buf.getInt();
        assertEquals(msg.length - 1, length);
    }

    @Test
    void testBuildQueryMessage() {
        byte[] msg = Protocol.buildQueryMessage("SELECT 1");
        assertEquals('Q', msg[0]);
        ByteBuffer buf = ByteBuffer.wrap(msg, 1, 4);
        int length = buf.getInt();
        assertEquals(msg.length - 1, length);
        // Check SQL is in the payload
        String payload = new String(msg, 5, msg.length - 5, StandardCharsets.UTF_8);
        assertTrue(payload.startsWith("SELECT 1"));
    }

    @Test
    void testBuildParseMessage() {
        byte[] msg = Protocol.buildParseMessage("SELECT $1", "");
        assertEquals('P', msg[0]);
        assertTrue(msg.length > 5);
    }

    @Test
    void testBuildBindMessage() {
        byte[] msg = Protocol.buildBindMessage(new String[]{"hello", "42"}, "", "");
        assertEquals('B', msg[0]);
        assertTrue(msg.length > 5);
    }

    @Test
    void testBuildBindMessageWithNull() {
        byte[] msg = Protocol.buildBindMessage(new String[]{Protocol.NULL_VALUE}, "", "");
        assertEquals('B', msg[0]);
        assertTrue(msg.length > 5);
    }

    @Test
    void testBuildDescribeMessage() {
        byte[] msg = Protocol.buildDescribeMessage((byte) 'P', "");
        assertEquals('D', msg[0]);
    }

    @Test
    void testBuildExecuteMessage() {
        byte[] msg = Protocol.buildExecuteMessage("", 0);
        assertEquals('E', msg[0]);
    }

    @Test
    void testBuildSyncMessage() {
        byte[] msg = Protocol.buildSyncMessage();
        assertEquals(5, msg.length);
        assertEquals('S', msg[0]);
    }

    @Test
    void testBuildTerminateMessage() {
        byte[] msg = Protocol.buildTerminateMessage();
        assertEquals(5, msg.length);
        assertEquals('X', msg[0]);
    }

    @Test
    void testBuildSASLInitialResponse() {
        byte[] clientFirst = "n,,n=user,r=nonce".getBytes(StandardCharsets.UTF_8);
        byte[] msg = Protocol.buildSASLInitialResponse("SCRAM-SHA-256", clientFirst);
        assertEquals('p', msg[0]);
        assertTrue(msg.length > 5);
    }

    @Test
    void testBuildSASLResponse() {
        byte[] clientFinal = "c=biws,r=nonce,p=proof".getBytes(StandardCharsets.UTF_8);
        byte[] msg = Protocol.buildSASLResponse(clientFinal);
        assertEquals('p', msg[0]);
    }

    @Test
    void testParseErrorFields() {
        // Build an error response: S\0ERROR\0C\023000\0M\0test error\0\0
        byte[] payload = "SERROR\0C23000\0Mtest error\0\0".getBytes(StandardCharsets.UTF_8);
        Map<Byte, String> fields = Protocol.parseErrorFields(payload);
        assertEquals("ERROR", fields.get((byte) 'S'));
        assertEquals("23000", fields.get((byte) 'C'));
        assertEquals("test error", fields.get((byte) 'M'));
    }

    @Test
    void testParseRowDescription() throws IOException {
        // Build a row description: 1 field, name="id", tableOID=0, colIdx=0, typeOID=23 (INT4), typeSize=4, typeMod=-1, fmtCode=0
        ByteBuffer buf = ByteBuffer.allocate(256);
        buf.putShort((short) 1); // field count
        buf.put("id\0".getBytes(StandardCharsets.UTF_8));
        buf.putInt(0); // tableOID
        buf.putShort((short) 0); // colIdx
        buf.putInt(23); // typeOID (INT4)
        buf.putShort((short) 4); // typeSize
        buf.putInt(-1); // typeMod
        buf.putShort((short) 0); // fmtCode
        buf.flip();
        byte[] payload = new byte[buf.remaining()];
        buf.get(payload);

        List<Protocol.FieldDescription> fields = Protocol.parseRowDescription(payload);
        assertEquals(1, fields.size());
        assertEquals("id", fields.get(0).name);
        assertEquals(23, fields.get(0).typeOID);
    }

    @Test
    void testParseDataRow() throws IOException {
        // 2 columns: "hello" and NULL
        ByteBuffer buf = ByteBuffer.allocate(256);
        buf.putShort((short) 2);
        byte[] val1 = "hello".getBytes(StandardCharsets.UTF_8);
        buf.putInt(val1.length);
        buf.put(val1);
        buf.putInt(-1); // NULL
        buf.flip();
        byte[] payload = new byte[buf.remaining()];
        buf.get(payload);

        List<byte[]> values = Protocol.parseDataRow(payload);
        assertEquals(2, values.size());
        assertEquals("hello", new String(values.get(0), StandardCharsets.UTF_8));
        assertNull(values.get(1));
    }

    @Test
    void testParseCString() {
        byte[] payload = "hello world\0extra".getBytes(StandardCharsets.UTF_8);
        assertEquals("hello world", Protocol.parseCString(payload));
    }

    @Test
    void testParseRowCount() {
        assertEquals(5, Protocol.parseRowCount("SELECT 5"));
        assertEquals(3, Protocol.parseRowCount("INSERT 0 3"));
        assertEquals(0, Protocol.parseRowCount("CREATE TABLE"));
    }

    @Test
    void testParseRowDescriptionMultipleFields() throws IOException {
        ByteBuffer buf = ByteBuffer.allocate(512);
        buf.putShort((short) 3); // 3 fields
        for (String name : new String[]{"id", "name", "age"}) {
            buf.put((name + "\0").getBytes(StandardCharsets.UTF_8));
            buf.putInt(0); // tableOID
            buf.putShort((short) 0); // colIdx
            buf.putInt(name.equals("id") ? 23 : (name.equals("name") ? 25 : 23)); // typeOID
            buf.putShort((short) 4); // typeSize
            buf.putInt(-1); // typeMod
            buf.putShort((short) 0); // fmtCode
        }
        buf.flip();
        byte[] payload = new byte[buf.remaining()];
        buf.get(payload);

        List<Protocol.FieldDescription> fields = Protocol.parseRowDescription(payload);
        assertEquals(3, fields.size());
        assertEquals("id", fields.get(0).name);
        assertEquals("name", fields.get(1).name);
        assertEquals("age", fields.get(2).name);
    }
}
